core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/uni/ForEachUniNode.java

```diff
@@ -57,44 +65,44 @@ public void retract(A a) {
         }
-        if (tuple.state.isDirty()) {
-            if (tuple.state == BavetTupleState.DYING || tuple.state == BavetTupleState.ABORTING) {
-                throw new IllegalStateException("The fact (" + a + ") was already retracted, so it cannot retract.");
-            }
-            tuple.state = BavetTupleState.ABORTING;
-        } else {
-            tuple.state = BavetTupleState.DYING;
-            dirtyTupleQueue.add(tuple);
+        switch (tuple.state) {
+            case CREATING:
+                // Kill it before it propagates
+                tuple.dirtyListEntry.remove();
+                tuple.state = BavetTupleState.ABORTING;
+                break;
+            case UPDATING:
+                tuple.dirtyListEntry.remove();
+                tuple.state = BavetTupleState.DYING;
+                tuple.dirtyListEntry = dirtyDyingList.add(tuple);
+                break;
+            case OK:
+                tuple.state = BavetTupleState.DYING;
+                tuple.dirtyListEntry = dirtyDyingList.add(tuple);
+                break;
+            case DYING:
+            case ABORTING:
+            case DEAD:
+            default:
+                throw new IllegalStateException("Impossible state: The tuple (" + tuple + ") in node (" +
+                        this + ") is in an unexpected state (" + tuple.state + ").");
         }
     }
 
     @Override
     public void calculateScore() {
-        for (UniTupleImpl<A> tuple : dirtyTupleQueue) {
-            switch (tuple.state) {
-                case CREATING:
-                    nextNodesTupleLifecycle.insert(tuple);
-                    tuple.state = BavetTupleState.OK;
-                    break;
-                case UPDATING:
-                    nextNodesTupleLifecycle.update(tuple);
-                    tuple.state = BavetTupleState.OK;
-                    break;
-                case DYING:
-                    nextNodesTupleLifecycle.retract(tuple);
-                    tuple.state = BavetTupleState.DEAD;
-                    break;
-                case ABORTING:
-                    tuple.state = BavetTupleState.DEAD;
-                    break;
-                case OK:
-                case DEAD:
-                default:
-                    throw new IllegalStateException("Impossible state: The tuple (" + tuple + ") in node (" +
-                            this + ") is in an unexpected state (" + tuple.state + ").");
-            }
-        }
-        dirtyTupleQueue.clear();
+        dirtyCreatingList.forEachAndClear(tuple -> {
+            nextNodesTupleLifecycle.insert(tuple);
+            tuple.state = BavetTupleState.OK;
+        });
+        dirtyUpdatingList.forEachAndClear(tuple -> {
+            nextNodesTupleLifecycle.update(tuple);
+            tuple.state = BavetTupleState.OK;
+        });
+        dirtyDyingList.forEachAndClear(tuple -> {
+            nextNodesTupleLifecycle.retract(tuple);
+            tuple.state = BavetTupleState.DEAD;
```

Comment:
This needs to be the exact oposite order, IMO.

First you retract, so that the joins become smaller first.
Then you update, to modify what's there already.
And finally you add new things.

That said, I doubt that this will improve performance, because what it is doing is adding a boatload of new TupleList entries.

Reply:
Good catch!